File: ios/App/AuthAPI.swift
import Foundation

struct OtpRequestPayload: Codable {
    let phoneNumber: String
}

struct OtpVerifyPayload: Codable {
    let phoneNumber: String
    let code: String
}

struct OtpVerifyResponse: Codable {
    let token: String
}

final class AuthAPI {
    private let baseURL: URL
    private let session: URLSession

    init(baseURL: URL = URL(string: ProcessInfo.processInfo.environment["API_BASE_URL"] ?? "https://example.com")!, session: URLSession = .shared) {
        self.baseURL = baseURL
        self.session = session
    }

    func requestOtp(phoneNumber: String) async throws {
        var request = URLRequest(url: baseURL.appending(path: "/api/mobile/whatsapp/request"))
        request.httpMethod = "POST"
        request.addValue("application/json", forHTTPHeaderField: "Content-Type")
        request.httpBody = try JSONEncoder().encode(OtpRequestPayload(phoneNumber: phoneNumber))
        let (_, response) = try await session.data(for: request)
        guard let httpResponse = response as? HTTPURLResponse, 200..<300 ~= httpResponse.statusCode else {
            throw URLError(.badServerResponse)
        }
    }

    func verifyOtp(phoneNumber: String, code: String) async throws -> String {
        var request = URLRequest(url: baseURL.appending(path: "/api/mobile/whatsapp/verify"))
        request.httpMethod = "POST"
        request.addValue("application/json", forHTTPHeaderField: "Content-Type")
        request.httpBody = try JSONEncoder().encode(OtpVerifyPayload(phoneNumber: phoneNumber, code: code))
        let (data, response) = try await session.data(for: request)
        guard let httpResponse = response as? HTTPURLResponse, 200..<300 ~= httpResponse.statusCode else {
            throw URLError(.badServerResponse)
        }
        let payload = try JSONDecoder().decode(OtpVerifyResponse.self, from: data)
        return payload.token
    }
}

File: android/app/src/main/java/com/rayon/mobile/network/AuthApi.kt
package com.rayon.mobile.network

import io.ktor.client.HttpClient
import io.ktor.client.call.body
import io.ktor.client.request.post
import io.ktor.client.request.setBody
import io.ktor.client.request.url
import retrofit2.http.Body
import retrofit2.http.POST

private const val DEFAULT_ENDPOINT = "api/mobile/whatsapp"

data class OtpRequest(val phoneNumber: String)
data class OtpVerifyRequest(val phoneNumber: String, val code: String)
data class OtpRequestResponse(val correlationId: String)
data class OtpVerifyResponse(val token: String)

interface WhatsAppAuthService {
  @POST("$DEFAULT_ENDPOINT/request")
  suspend fun request(@Body payload: OtpRequest): OtpRequestResponse

  @POST("$DEFAULT_ENDPOINT/verify")
  suspend fun verify(@Body payload: OtpVerifyRequest): OtpVerifyResponse
}

class SupabaseWhatsAppClient(private val httpClient: HttpClient, private val baseUrl: String) {
  suspend fun requestOtp(payload: OtpRequest): OtpRequestResponse =
    httpClient.post {
      url("$baseUrl/$DEFAULT_ENDPOINT/request")
      setBody(payload)
    }.body()

  suspend fun verifyOtp(payload: OtpVerifyRequest): OtpVerifyResponse =
    httpClient.post {
      url("$baseUrl/$DEFAULT_ENDPOINT/verify")
      setBody(payload)
    }.body()
}
